transports/quic/src/hole_punching.rs

```diff
@@ -0,0 +1,126 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{
+    channel::oneshot,
+    future::{Fuse, FusedFuture},
+    prelude::*,
+};
+use futures_timer::Delay;
+use libp2p_identity::PeerId;
+use rand::{distributions, Rng};
+
+use crate::{
+    endpoint::{self, ToEndpoint},
+    Connecting, Connection, Error,
+};
+
+pub(crate) type HolePunchMap =
+    Arc<Mutex<HashMap<(SocketAddr, PeerId), oneshot::Sender<(PeerId, Connection)>>>>;
+
+pub(crate) struct MaybeHolePunchedConnection {
+    hole_punch_map: HolePunchMap,
+    addr: SocketAddr,
+    upgrade: Fuse<Connecting>,
+}
+
+impl MaybeHolePunchedConnection {
+    pub(crate) fn new(hole_punch_map: HolePunchMap, addr: SocketAddr, upgrade: Connecting) -> Self {
+        Self {
+            hole_punch_map,
+            addr,
+            upgrade: upgrade.fuse(),
+        }
+    }
+}
+
+impl Future for MaybeHolePunchedConnection {
+    type Output = Result<(PeerId, Connection), Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let (peer_id, connection) = futures::ready!(self.upgrade.poll_unpin(cx))?;
+        let addr = self.addr;
+        let mut hole_punch_map = self.hole_punch_map.lock().unwrap();
+        if let Some(sender) = hole_punch_map.remove(&(addr, peer_id)) {
+            if let Err(connection) = sender.send((peer_id, connection)) {
+                Poll::Ready(Ok(connection))
+            } else {
+                Poll::Pending
```

Comment:
In the case where we successfully send the hole punched connection via `sender.send`, we will indefinitely return `Poll::Pending` from the `MaybeHolePunchedConnection` `Future`, right?

That might not be ideal, as we now depend on some upper layer timeout to eventually clean up the indefinitely pending `Future`.

Off the top of my head I would suggest returning a `Poll::Ready(Err(_))` to make sure the `Future` terminates. Open for other suggestions.

Reply:
What should be the error type if we do return `Poll::Ready(Err(_))`?

Any other suggestions @kpp, @thomaseizinger?

Reply:
Another thing to note is that returning an `Err` will result in a [`ListenFailure` event being generated](https://github.com/libp2p/rust-libp2p/blob/75edcfcdb09f5d8f55b7347e90e62203d42bc357/swarm/src/lib.rs#L783-L805).